### ISSUE TITLE
Confirm change for model with field m2m

### ISSUE THREAD

**Issue by IRemsha** [TrangPham/django-admin-confirm#8]
When I try save change for models where there m2m field I get TypeError `object of type 'ManyRelatedManager' has no len()`

`  File "/usr/local/lib/python3.7/site-packages/django/forms/models.py", line 1385, in has_changed
    if len(initial) != len(data):
`

**Comment by TrangPham** [comment]
Hi IRemsha,

Thanks for raising this issue. I had not tested it against models with inlines/m2m relations. 

If you wish to tackle the problem, you can repro the issue within the test project, address it and add unit tests and I can review. 

Otherwise, I will take a look into this when I have some time. 

Thanks!

**Comment by TrangPham** [comment]
Hi 

I tried to reproduce the issue you are seeing and was not able to. 

Could you please paste what your models and admin look like? And what versions of Django you're using?

Here are the two test cases that I tried:

**Test case 1**
```python
class InventoryInline(admin.TabularInline):
    model = Inventory

class ShopAdmin(AdminConfirmMixin, admin.ModelAdmin):
    confirmation_fields = ["name"]
    confirm_change = True
    inlines = [InventoryInline]
    actions = ["show_message", "show_message_no_confirmation"]

    @confirm_action
    def show_message(modeladmin, request, queryset):
        shops = ", ".join(shop.name for shop in queryset)
        modeladmin.message_user(request, f"You selected with confirmation: {shops}")

    show_message.allowed_permissions = ("delete",)

    def show_message_no_confirmation(modeladmin, request, queryset):
        shops = ", ".join(shop.name for shop in queryset)
        modeladmin.message_user(request, f"You selected without confirmation: {shops}")

    def has_delete_permission(self, request, obj=None):
        return request.user.is_superuser

```
Showed no issue saving the shop model and no issues confirming the changes on the name field.


**Test case 2**
```python
class InventoryInline(AdminConfirmMixin, admin.TabularInline):  
    # FYI: AdminConfirmMixin does not work with Inlines and has not been designed to, but it could be a feature in the future
    model = Inventory
    confirm_change = True
    confirm_add = True
    confirmation_fields = ["quantity"]


class ShopAdmin(AdminConfirmMixin, admin.ModelAdmin):
    confirmation_fields = ["name"]
    confirm_change = True
    confirm_add = True
    inlines = [InventoryInline]
    actions = ["show_message", "show_message_no_confirmation"]

    @confirm_action
    def show_message(modeladmin, request, queryset):
        shops = ", ".join(shop.name for shop in queryset)
        modeladmin.message_user(request, f"You selected with confirmation: {shops}")

    show_message.allowed_permissions = ("delete",)

    def show_message_no_confirmation(modeladmin, request, queryset):
        shops = ", ".join(shop.name for shop in queryset)
        modeladmin.message_user(request, f"You selected without confirmation: {shops}")

    def has_delete_permission(self, request, obj=None):
        return request.user.is_superuser
```
Showed no problems with saving new instances with confirm_add for shop, showed no issue for confirm_change for shop.
**DOES NOT** confirm the changes on the inlines on add or change - because this is not a feature (yet?)

**Comment by TrangPham** [comment]
Ah oops! I misunderstood... I think you meant `models.ManyToManyField` :P I will try to reproduce again

**Comment by MichaelSwartz** [comment]
@TrangPham Thanks for this library! I was actually about to post an issue requesting support for Inlines, like test case 2 above. I unfortunately don't have the expertise to do this myself but would be very appreciative if someone else builds it. Totally understand that it's probably not a priority for you now.

Thanks!

**Comment by TrangPham** [comment]
@IRemsha I was able to reproduce the issue and actually fixed a couple more related to m2m fields. Thanks so much for bringing this to our attention

@MichaelSwartz Thanks :) I have added it to the potential future features list. I think it makes a lot of sense to have that and will look into it after adding a full CI test suite. I would also be happy to help advise and review code if you wanted to try tackling it yourself.

**Comment by TrangPham** [comment]
@IRemsha  Could I bother you to test using the new version before I release it? To do that you can install via:
```
python -m pip install --index-url https://test.pypi.org/simple/ django_admin_confirm==0.2.3
```

**Comment by IRemsha** [comment]
@TrangPham I get this error, after set version to `0.2.3`. Do you have some idea? 

ERROR:
`django.contrib.gis.geos.error.GEOSException: Could not parse version info string "3.7.1-CAPI-1.11.1 27a5e771"`

Pipfile:
```
[[source]]
name = "pypi"
url = "https://pypi.org/simple"
verify_ssl = true

[[source]]
name= "test-pypip"
url = "https://test.pypi.org/simple"
verify_ssl = true

[dev-packages]

[packages]
django = "~=3.1"
django-storages = "~=1.10"
boto3 = "~=1.15"
django-environ = "~=0.4"
gunicorn = "~=20.0"
postgis = "~=1.0"
psycopg2 = "~=2.8"
pillow = "~=7.2"
django-cors-headers = "~=3.5"
pyjwt = "~=1.7"
pyyaml = "~=5.3"
flake8 = "~=3.8"
django-cassandra-engine = "~=1.6"
channels = "~=2.4"
channels-redis = "~=3.1"
responses = "~=0.12"
factory-boy = "~=3.0"
pyfcm = "~=1.4"
requests = "~=2.24"
daphne = "~=2.5"
django-modeladmin-reorder = "~=0.3"
django-admin-confirm = {version="*", index="test-pypip"}

[requires]
python_version = "3.8"
```

**Comment by TrangPham** [comment]
Hmm I'm not sure what exactly is going on there. This library doesn't use any geo data processing. Doing a quick google, I see some workarounds you could try: https://stackoverflow.com/questions/18643998/geodjango-geosexception-error

I don't think running python 3.8.5 is part of the issue.

**Comment by IRemsha** [comment]
Problems resolved if use Django 2.*

**Comment by IRemsha** [comment]
Now I get field  `<QuerySet [<City: Mastery Kenobi>]>` maybe need unzip QuerySet for beauty?




And

**Comment by TrangPham** [comment]
Hey IRemsha,

that’s what you’re supposed to get currently :P I did not go the extra mile to parse the querysets to nice text. But I’m glad the fix is working. I’ll see if I have time to do that soon to include it in the release. At least for now you should be unblocked for confirming m2m fields even though it’s not formatted nicely. 

The geos version issue is probably unrelated to this library. But you could check out the issues on postgis to see if others have issues with Django 3.

**Comment by TrangPham** [comment]
I was trying to test this further and learnt that `QueryDict.items()` would only return the last item of a list. 

https://docs.djangoproject.com/en/3.1/ref/request-response/#django.http.QueryDict.items

So that means that we still don't support m2m fields because confirming them would change the past in value... 😞 I will have to find a way around this...

**Comment by TrangPham** [comment]
Okay 🙆 I think this should fix the issue that I found with not saving all the values in m2m fields that were selected. 

```
python -m pip install --index-url https://test.pypi.org/simple/ django_admin_confirm==0.2.3.dev1
```

**Comment by IRemsha** [comment]
@TrangPham 
Hi, I get error 
```
TemplateSyntaxError at /admin/api/community/fff741e8-31aa-4367-8726-0850f4c809b5/change/
'formatting' is not a registered tag library. Must be one of:
admin_list
admin_modify
admin_static
admin_urls
cache
i18n
l10n
log
static
staticfiles
tz
```

When I change Model with any field.

I use django_admin_confirm==0.2.3.dev1 with Django 2.2

**Comment by TrangPham** [comment]
@IRemsha Sorry about that! 🤦 I tested and got that too. I have now added more unit tests, CI testing + coveralls, and updated the formatting of the m2m fields. I tested saving the m2m fields and added unit tests for them too.

EXAMPLE
<img width="1467" alt="Screen Shot 2021-02-20 at 8 45 52 AM" src="https://user-images.githubusercontent.com/2117313/108602986-1ca8af80-735a-11eb-953e-a25273c16fa2.png">

This would take the `str()` value of the model, so you can override `def __str__` on your models to affect it and other places this model is shown by django admin.

I uploaded another dev version `0.2.3.dev4`

Thanks again so much for helping confirm the fixes.

**Comment by IRemsha** [comment]
@TrangPham Thank for job!
M2m field work correctly and beauty.

I want warn you be careful with FileField. I think in current value we'll see bytearray and None in New value. I think the most correct way is not to check these fields at all. 

Btw, is it possible now to exclude the fields from the change confirmation page now?

**Comment by TrangPham** [comment]
What do you by exclude? 

If you mean not check if the value changed then setting `confirmation_fields` would exclude any field not specified.

If you mean exclude in that it doesn't touch it at all....

Unfortunately :( the way that the Mixin works currently may modify the field value, even if it's not being checked for changes.
I'm not sure if `v0.2.2` modifies the data less...

https://github.com/TrangPham/django-admin-confirm/blob/c381eb4b87136762cc54991304d32a15c311e4c5/admin_confirm/admin.py#L169-L178

VS the new code...

https://github.com/TrangPham/django-admin-confirm/blob/fea907b3c8b68e5c164b59c48110a837738ea0a0/admin_confirm/admin.py#L140-L156 

This is because in the background it passes the POST params from the original "Save" click to the confirmation page, and the confirmation page puts this into a hidden form. I'm not sure if there's another better way of passing these. I can look into that.

I will open up another issue about the other special field types - and then have the code check for supported field types and fail hard on unsupported fields.

I should put a huge warning on the readme about this before I release any change.

**Comment by IRemsha** [comment]
@TrangPham Hi, in dev version bug appeared

It's work in version 2.2.3
![Снимок экрана 2021-02-21 111351](https://user-images.githubusercontent.com/40166328/108617265-af1e8100-7436-11eb-8157-74c5a87ee36b.png)

But don't wok in version 2.2.3.dev4
![Снимок экрана 2021-02-21 111756](https://user-images.githubusercontent.com/40166328/108617290-decd8900-7436-11eb-934d-0a73ff30018b.png)

**Comment by IRemsha** [comment]
> @TrangPham Hi, in dev version bug appeared
> I use ImageField. With FileField most likely works in a similar way
> 
> It's work in version 2.2.3
> ![Снимок экрана 2021-02-21 111351](https://user-images.githubusercontent.com/40166328/108617265-af1e8100-7436-11eb-8157-74c5a87ee36b.png)
> 
> But don't wok in version 2.2.3.dev4
> ![Снимок экрана 2021-02-21 111756](https://user-images.githubusercontent.com/40166328/108617290-decd8900-7436-11eb-934d-0a73ff30018b.png)

**Comment by TrangPham** [comment]
@IRemsha 

Thanks! That info is really helpful. 

I'll add in some test cases for FileField and ImageField. Are there any other field types that might need special treatment?

**Comment by IRemsha** [comment]
At the moment I have not found such fields

**Comment by TrangPham** [comment]
Thanks so much @IRemsha for helping to test these things. 

I have uploaded another dev release version `0.2.3.dev5` on test-pypi. You're welcome to give this a try. 

If you want to know the difference between this dev version vs the ones before, there's a good summary on #16.

I have not added the tests for the FileField and ImageField yet.

**Comment by IRemsha** [comment]
Hi, `fix(MR-16)` added to version `0.2.3.dev5`?

**Comment by TrangPham** [comment]
Hi @IRemsha - I was working really late last night. `fix(MR-16)` should fix it I think. I added some tests and manually tested a bit as well. But it's not in the `dev5` version. I will upload another dev release now

**Comment by TrangPham** [comment]
I have uploaded to testpypi `0.2.3.dev6` and did a couple manual tests after I installed from testpypi. Let me know if it works for you. 

😿 I tried probably 5-8 different solutions within the space of hours. Hoping this last soln works, but honestly there's probably some really uncommon edge cases that it doesn't work for.

**Comment by IRemsha** [comment]
Hi! I founded some errors:
1. For models with field m2m
```
Direct assignment to the forward side of a many-to-many set is prohibited. Use cities.set() instead. 
```
(cities this m2m field in my model)

2.  For models with custom fields
```
'Community' object has no attribute 'preview_logo'
```
Community  has no `preview_logo` it's method created in `admin.py` and appended to `fields` and `fieds_readonly`

3. For models where set Form
```
'EducationFormFormSet' object has no attribute 'new_objects'
```

**Comment by TrangPham** [comment]
> Hi! I found some errors:
> 
> 1. For models with field m2m
> 
> ```
> Direct assignment to the forward side of a many-to-many set is prohibited. Use cities.set() instead
> ```
> 
> (cities this m2m field in my model)

For this you do you have something like:

```
class City(Model):
    ...

class World(Model):
    cities = ManyToManyField(City, on_delete= ...)

class WorldAdmin(AdminConfirmMixin, ModelAdmin):
     confirm_change = True

admin.site.register(World, WorldAdmin)
         
```

> 
> 1. For models with custom fields
> 
> ```
> 'Community' object has no attribute 'preview_logo'
> ```
> 
> Community has no `preview_logo` it's method created in `admin.py` and appended to `fields` and `fieds_readonly`
> 

Do you mean `readonly_fields`?

> 1. For models where set Form
> 
> ```
> 'EducationFormFormSet' object has no attribute 'new_objects'
> ```

For this ModelAdmin, how is the form defined on admin? Is it:

```
class MyModelAdmin(ModelAdmin):
    form = EducationForm
```

**Comment by IRemsha** [comment]
1. Yes, that's what it says. Prior to the dv6 version, there were no errors in this location
2. Yes, error in line 188, file admin.py
3.  The only place where I use the EducationForm I created is view.py for data validation

**Comment by TrangPham** [comment]
Hey @IRemsha,

I went back to the drawing board and tried a different strategy for handling the ImageField and FileField. 
`0.2.3.dev7` is on testpypi now. 

I feel bad for having you try so many versions, and really appreciate your help! It's tricky to create unit tests for all the different variations of ModelAdmin that there can be.

**Comment by IRemsha** [comment]
Hi @TrangPham, 
My cases started working. If I find errors in the future I will try to report them as informatively as possible
Thank you so much for your work! you are an amazing author because you respond quickly to user requests!

**Comment by IRemsha** [comment]
Hmm, In one model I get error
```
ManagementForm data is missing or has been tampered with'
```
But I still don't understand why

**Comment by TrangPham** [comment]
> Hmm, In one model I get error
> 
> ```
> ManagementForm data is missing or has been tampered with'
> ```
> 
> But I still don't understand why

Are you getting this with ModelAdmin with inlines or where the `form` or `get_form` is set? Or is it happening with regular/simple admin pages as well?

**Comment by IRemsha** [comment]
Well, I get very strange error
```
AttributeError at /admin/api/apiuser/a7a38d4e-5772-4756-9744-3f467dfdb620/change/
'Community' object has no attribute 'photo'
```
I changed model `Apiuser`, but get error for `Community`

**Comment by TrangPham** [comment]
> Well, I get very strange error
> 
> ```
> AttributeError at /admin/api/apiuser/a7a38d4e-5772-4756-9744-3f467dfdb620/change/
> 'Community' object has no attribute 'photo'
> ```
> 
> I changed model `Apiuser`, but get error for `Community`
> Before that, I edited `Community`

I know why this one is happening - I have a fix for it and will put out another version, maybe tomorrow

**Comment by TrangPham** [comment]
I have addressed the two issues you were having recently. 

`0.2.3.dev9` is the new version on testpypi. If it works then I will release an alpha to pypi.

---
**Problem**: `ManagementForm data is missing or has been tampered with'`
**Root Cause**: Not passing formset data after confirmation
**Test Method**: Manual with one inline set
**Todo**: Test with multiple inlines

---
**Problem**: 
```
AttributeError at /admin/api/apiuser/a7a38d4e-5772-4756-9744-3f467dfdb620/change/
'Community' object has no attribute 'photo'
```
**Root Cause**: Old cache not cleared for another model. Caching is used for ModelAdmin with multipart forms (aka files and images). 
**Solution**: 
- Clear cache where it makes sense or is not required
- Add timeout to cache, `ADMIN_CONFIRM_CACHE_TIMEOUT` is the setting for this if you want to set it to something else in your Django Settings file. Default is 10 seconds.
- Make cache control private
**Test Method**: Unit tests
**Todo:** Test to ensure that cache is private when multiple users simultaneously using admin

**Comment by IRemsha** [comment]
@TrangPham 
Hi, managed to solve the problem with inlines?
`ManagementForm data is missing or has been tampered with`

**Comment by TrangPham** [comment]
@IRemsha Ah sorry yes that's what I was saying with last post - I think I fixed those two issues. You can try `0.2.3.dev9` from test pypi

**Comment by TrangPham** [comment]
:) I have added integration tests for confirmation of models with inlines on the page, and models with files. 
Unit tests for the m2m fields are also in place. 

New version: `0.2.3a0` on testpypi. And I will probably do a pre-release on pypi soon

**Comment by IRemsha** [comment]
@TrangPham , HI! How does uploading images after the confirmation form work now?

My images don't load on S3 if the confirmation form was called

**Comment by TrangPham** [comment]
Oh I have not tested it with S3. I was using local storage to test it. I
will have to look into how the storage internals for Django work and
get back to you. For now I would say, don’t use confirmation on admin forms
with uploads/ files on them.

**Comment by TrangPham** [comment]
> Issue: ManyToManyField causes `object of type 'ManyRelatedManager' has no len()` on confirmation 
Affected version: 0.2.2
Fixed in: 0.2.3a0
